### PR TITLE
feat: [Sites-25915] API spec for api key management

### PR DESCRIPTION
**Pull request by swetabar**
Proposing 3 new endpoints:
* `POST /tools/api-keys` - Create a new API Key
* `GET /tools/api-keys` - Retrieve the API Key metadata for all API Keys created by the user
* `DELETE /tools/api-keys/{apiKeyId}` - Delete the API with the Id: apiKeyId 

## Related Issues
https://jira.corp.adobe.com/browse/SITES-25915
Wiki: https://wiki.corp.adobe.com/display/AEMSites/API+Key+Management

### CODE REVIEW

File: docs/openapi/parameters.yaml [hunk 1]
```diff
@@ -119,3 +119,10 @@ jobId:
   required: true
   schema:
     $ref: './schemas.yaml#/Id'
+apiKeyId:
+    name: apiKeyId
+    description: The API key ID in uuid format
+    in: path
+    required: true
+    schema:
+        $ref: './schemas.yaml#/Id'
```
Comment (blefebvre) [inline docs/openapi/parameters.yaml:128]: Indentation looks off here.

File: docs/openapi/responses.yaml [hunk 2]
```diff
@@ -96,6 +96,12 @@
       $ref: './headers.yaml#/xError'
 '401':
   description: Not authenticated.
+'403-max-api-keys-exceeded':
+  description: |
+    Forbidden. The maximum number of API keys that can be created has been exceeded.
```
Comment (blefebvre) [inline docs/openapi/responses.yaml:101]:
```suggestion
    Forbidden. The maximum number of API keys that can be created per client has been exceeded.
```

File: docs/openapi/schemas.yaml [hunk 3]
```diff
@@ -1517,3 +1520,91 @@ ImportAssistantResponse:
...
+      description: The name of the API key
+      type: string
+    apiKey:
+      description: The API key
+      type: string
+    scopes:
+      description: The scopes of the API key
+      type: array
+      items:
+        type: string
```
Comment (blefebvre) [inline docs/openapi/schemas.yaml:1562]: Will this be an `object` type, as they are currently stored in the DB?

File: docs/openapi/schemas.yaml [hunk 4]
```diff
@@ -1517,3 +1520,91 @@ ImportAssistantResponse:
...
+    domains:
+      description: The domains the API key is allowed to access
```
Comment (blefebvre) [inline docs/openapi/schemas.yaml:1564]: If the above `scopes` is an array of objects, then `imports.write` (or any other scope which needs it) will contain the domains information, so it wouldn't need to be duplicated here.

File: docs/openapi/schemas.yaml [hunk 5]
```diff
@@ -1517,3 +1520,91 @@ ImportAssistantResponse:
...
+      $ref: '#/ImsOrganizationId'
+    imsUserId:
+      description: The IMS user ID
+      $ref: '#/ImsUserId'
+    createdAt:
+      description: The date and time when the API key was created
+      $ref: '#/DateTime'
+    expiresAt:
+      description: The date and time when the API key expires
+      $ref: '#/DateTime'
```
Comment (blefebvre) [inline docs/openapi/schemas.yaml:1579]: Do we need a `deletedAt` here as well, to handle when the user initiates a delete?

File: docs/openapi/schemas.yaml [hunk 6]
```diff
@@ -1517,3 +1520,91 @@ ImportAssistantResponse:
...
+ApiKeyMetadataResponse:
```
Comment (blefebvre) [inline docs/openapi/schemas.yaml:1580]: Does this type just hide the `apiKey` value? Any way we could accomplish this without the duplication?

File: docs/openapi/tools-api.yaml [hunk 7]
```diff
@@ -218,3 +218,115 @@ import-assistant:
...
+api-keys:
+  post:
+    tags:
+      - api-keys
+    summary: Create a new API key
+    description: |
+      Note: Not implemented yet.
+      
+      This endpoint can be used to create a new API key with a set of scopes. The scoped API key will be
+      returned in the response only once. The scoped API key can be used to authenticate the spacecat APIs by
```
Comment (blefebvre) [inline docs/openapi/tools-api.yaml:230]:
```suggestion
      returned in the response only once. The scoped API key can be used to authenticate the Spacecat APIs by 
```

File: docs/openapi/tools-api.yaml [hunk 8]
```diff
@@ -218,3 +218,115 @@ import-assistant:
...
+      '401':
+        $ref: './responses.yaml#/401'
+      '403':
+        $ref: './responses.yaml#/403-max-api-keys-exceeded'
+      '500':
+        $ref: './responses.yaml#/500'
+  get:
+    tags:
+      - api-keys
+    summary: Get all API keys
```
Comment (blefebvre) [inline docs/openapi/tools-api.yaml:265]:
```suggestion
    summary: Get all API keys owned by the given IMS client
```

File: docs/openapi/tools-api.yaml [hunk 9]
```diff
@@ -218,3 +218,115 @@ import-assistant:
...
+    description: |
+      Note: Not implemented yet.
+      
+      This endpoint can be used to get all the API keys created by the client.
```
Comment (blefebvre) [inline docs/openapi/tools-api.yaml:269]:
```suggestion
      This endpoint can be used to list all the API keys created by the client.
```

File: docs/openapi/tools-api.yaml [hunk 10]
```diff
@@ -218,3 +218,115 @@ import-assistant:
...
+api-key-deletion:
+  delete:
+    tags:
+      - api-keys
+    summary: Delete an API key
+    description: |
+      Note: Not implemented yet.
+      
+      This endpoint can be used to delete an API key. The API key will be deleted from the system and
+      will no longer be valid for authentication.
```
Comment (blefebvre) [inline docs/openapi/tools-api.yaml:304]:
```suggestion
      will no longer be valid for authentication. Only keys created by this IMS client can be deleted by it.
```

File: docs/openapi/tools-api.yaml [hunk 11]
```diff
@@ -218,3 +218,115 @@ import-assistant:
...
+      - ims_key: [ ]
+    parameters:
+      - $ref: './parameters.yaml#/apiKeyId'
+      - in: header
+        name: x-gw-ims-org-id
+        schema:
+          type: string
+        description: |
+          IMS Organization Id of the client.
+      - in: header
```
Comment (blefebvre) [inline docs/openapi/tools-api.yaml:316]: Is this necessary if we will receive the `apiKeyId` in the path?

File: docs/openapi/schemas.yaml (hunk 3 above)
Comment (swetabar) [reply docs/openapi/schemas.yaml:1562]: They will be stored as an array of strings. Not as a map object as we are moving domains to a higher level.

Comment (blefebvre) [reply docs/openapi/schemas.yaml:1562]: Will this make it trickier (more error prone) to serialize from the api-keys table to this response DTO?

I totally see the benefit of the higher level `features` prop and specifying domains at the top level of the "create key" request, but wonder if that should be considered a convenience helper that we don't need to mirror when we actually respond back with the key. After all we will be losing the `features` property anyways, in favour of specifying the exact scopes that a particular feature (or set of features) maps to.

File: docs/openapi/tools-api.yaml (hunk 11 above)
Comment (swetabar) [reply docs/openapi/tools-api.yaml:316]: Just to confirm that the apiKeyId requested is created by the user. Also, for simplicity - keeping the same header for all apiKey related requests.

File: docs/openapi/schemas.yaml (hunk 5 above)
Comment (swetabar) [reply docs/openapi/schemas.yaml:1579]: Yes - can add a deletedAt, should work.

File: docs/openapi/schemas.yaml (hunk 3 above)
Comment (swetabar) [reply docs/openapi/schemas.yaml:1562]: I was thinking of modifying the ApiKeys table as well to make it more simplistic. Keep domains at a higher level. I'm open to re-think it.

File: docs/openapi/schemas.yaml (hunk 4 above)
Comment (swetabar) [reply docs/openapi/schemas.yaml:1564]: I was thinking about it - should we keep it consistent everywhere? Move `domains` to `scopes` everywhere. Or move `domains` to a higher level everywhere (even in the database).

File: docs/openapi/schemas.yaml (hunk 5 above)
Comment (swetabar) [reply docs/openapi/schemas.yaml:1579]: Should deletedAt be part of the ApiKeyMetadataResponse? As `ApiKeyResponse` is shown only once - when the user creates an API Key.

File: docs/openapi/api.yaml [hunk 12]
```diff
@@ -41,6 +41,8 @@ tags:
     description: Webhooks for receiving events
   - name: slack
     description: Slack channel management
+  - name: api-keys
+    description: API key management
```
Comment (atopper) [inline docs/openapi/api.yaml:45]: This seems to refer (eventually) to a description of a single api-key.  Was this internationally plural (`api-keys` vs `api-key`)?

File: docs/openapi/schemas.yaml [hunk 13]
```diff
@@ -40,6 +40,9 @@ Domain:
 ImsOrganizationId:
   type: string
   example: '1234567890ABCDEF12345678@AdobeOrg'
+ImsUserId:
+  type: string
+  example: '8G5129DA63CE4CD98A495EAD@8053114e63a16b59485e3c.e'
```
Comment (atopper) [inline docs/openapi/schemas.yaml:45]: This looks like a legit id.  Maybe use than obviously fake pattern to show an example: `123abc456def789fgh@09876543e12a345e6.e`

File: docs/openapi/api.yaml (hunk 12 above)
Comment (swetabar) [reply docs/openapi/api.yaml:45]: I kept it as api-keys as we could use the same endpoint: /tools/api-keys for both post and get.
`GET /tools/api-keys` retrieves all API keys created by a client.

File: docs/openapi/schemas.yaml (hunk 13 above)
Comment (swetabar) [reply docs/openapi/schemas.yaml:45]: Good point. I took a real one and modified the letters and alphabets.

File: docs/openapi/parameters.yaml (hunk 1 above)
Comment (swetabar) [reply docs/openapi/parameters.yaml]: Done.

File: docs/openapi/schemas.yaml (hunk 6 above)
Comment (swetabar) [reply docs/openapi/schemas.yaml:1580]: Something like this:
```
ApiKeyResponse:
  type: object
  properties:
    id:
      description: The ID of the API key
      $ref: '#/Id'
    name:
      description: The name of the API key
      type: string
    apiKey:
      description: The API key
      type: string
    scopes:
      description: The scopes of the API key
      type: array
      items:
        type: string
    domains:
      description: The domains the API key is allowed to access
      type: array
      items:
        type: string
    imsOrgId:
      description: The IMS organization ID
      $ref: '#/ImsOrganizationId'
    imsUserId:
      description: The IMS user ID
      $ref: '#/ImsUserId'
    createdAt:
      description: The date and time when the API key was created
      $ref: '#/DateTime'
    expiresAt:
      description: The date and time when the API key expires
      $ref: '#/DateTime'

ApiKeyMetadataResponse:
  allOf:
    - $ref: '#/ApiKeyResponse'
    - type: object
      properties:
        apiKey: {}
      additionalProperties: false
```

Comment (swetabar) [reply docs/openapi/schemas.yaml:1580]: But, in this case, it shows the apiKey as null.
Screenshot 2024-10-24 at 10 44 26 AM

File: docs/openapi/schemas.yaml (hunk 3 above)
Comment (blefebvre) [reply docs/openapi/schemas.yaml:1562]: As discussed at standup, let's mirror how we've modelled scopes in the api-keys table.

File: docs/openapi/schemas.yaml (hunk 5 above)
Comment (swetabar) [reply docs/openapi/schemas.yaml:1579]: We list the expired key as part of the APIKeyMetadataResponse, but not the deleted keys.

File: docs/openapi/tools-api.yaml (hunk 11 above)
Comment (swetabar) [reply docs/openapi/tools-api.yaml:316]: `x-api-key` needs to be removed.

File: docs/openapi/schemas.yaml (hunk 5 above)
Comment (swetabar) [reply docs/openapi/schemas.yaml:1579]: `deletedAt` will be an attribute in the ApiKeys table. We do not need to show it in the metadata response.

File: docs/openapi/schemas.yaml (hunk 6 above)
Comment (swetabar) [reply docs/openapi/schemas.yaml:1551]: Done.

File: docs/openapi/schemas.yaml [hunk 14]
```diff
@@ -37,6 +37,14 @@ Domain:
...
   format: domain
   example: 'www.example.com'
+ScopeName:
+  type: string
+  enum:
+    - 'imports.write'
+    - 'imports.read'
+    - 'imports.all_domains'
+    - 'imports.read_all'
+    - 'imports.delete'
```
Comment (blefebvre) [inline docs/openapi/schemas.yaml:47]: Do we need to call out these 3 in the docs?

File: docs/openapi/schemas.yaml [hunk 15]
```diff
@@ -1517,3 +1525,62 @@ ImportAssistantResponse:
     choices:
       - message:
           content: '.breadcrumbs'
+ApiKeyRequest:
+  type: object
+  required:
+    - name
+    - feature
```
Comment (blefebvre) [inline docs/openapi/schemas.yaml:1532]: Should `feature` be `features` instead, and an array? Thinking ahead to future use cases.

File: docs/openapi/schemas.yaml [hunk 16]
```diff
@@ -1517,3 +1525,62 @@ ImportAssistantResponse:
...
+  properties:
+    id:
+      description: The ID of the API key
+      $ref: '#/Id'
+    name:
+      description: The name of the API key
+      type: string
+    scopes:
+      description: The scopes of the API key
+      type: object
```
Comment (blefebvre) [inline docs/openapi/schemas.yaml:1562]: Believe this will be an array of item type `object`.

File: docs/openapi/tools-api.yaml [hunk 17]
```diff
@@ -218,3 +218,109 @@ import-assistant:
...
+    parameters:
+      - in: header
+        name: x-gw-ims-org-id
+        schema:
+          type: string
+        description: |
+          IMS Organization Id of the client.
+    responses:
+      '200':
+        description: API keys returned successfully.
```
Comment (blefebvre) [inline docs/openapi/tools-api.yaml:282]:
```suggestion
        description: API key metadata returned successfully. Note that the API key itself will not be included in this list, as it's only displayed once (at creation).
```

File: docs/openapi/tools-api.yaml [hunk 18]
```diff
@@ -218,3 +218,109 @@ import-assistant:
...
+        $ref: './responses.yaml#/500'
+api-key-deletion:
+  delete:
+    tags:
+      - api-keys
+    summary: Delete an API key
+    description: |
+      Note: Not implemented yet.
+      
+      This endpoint can be used to delete an API key. The API key will be deleted from the system and
```
Comment (blefebvre) [inline docs/openapi/tools-api.yaml:303]:
```suggestion
      This endpoint can be used to delete an API key. The API key will be marked as deleted in the system and
```

File: docs/openapi/tools-api.yaml [hunk 19]
```diff
@@ -218,3 +218,109 @@ import-assistant:
...
+      will no longer be valid for authentication. Only keys created by this IMS client can be deleted by it.
+    operationId: deleteApiKey
+    security:
+      - ims_key: [ ]
+    parameters:
+      - $ref: './parameters.yaml#/apiKeyId'
+      - in: header
+        name: x-gw-ims-org-id
+        schema:
```
Comment (blefebvre) [inline docs/openapi/tools-api.yaml:312]: Can this be reused? 
```
$ref: '#/ImsOrganizationId'
```

**Review by blefebvre — APPROVED** [review]
Few comments for your consideration 👍

File: docs/openapi/schemas.yaml (hunk 14 above)
Comment (swetabar) [reply docs/openapi/schemas.yaml:47]: I could remove it. As it's only for special users.

File: docs/openapi/schemas.yaml (hunk 15 above)
Comment (swetabar) [reply docs/openapi/schemas.yaml:1532]: Do we want to have the same api-key for imports and organizations? Maybe yes. I could change it to an array.

File: docs/openapi/tools-api.yaml (hunk 19 above)
Comment (swetabar) [reply docs/openapi/tools-api.yaml:312]: Keeping the orgId here as is, as `$ref: '#/ImsOrganizationId' ` is not accepting the name: `x-gw-ims-org-id`

Comment (blefebvre) [reply docs/openapi/tools-api.yaml:312]: Ah, OK. Can you add an example value in that case?

Comment (swetabar) [reply docs/openapi/tools-api.yaml:314]: Done.

**Review by solaris007 — APPROVED** [review]

**Pull request merged**